Create the working directory on container creation

coolljt0725 opened the pull request:
If create a container with -w to specify the working directory and
the directory does not exist in the container rootfs, the directory
will be created until the container start. It makes `docker export` of
a created container and a running container inconsistent. We always
use `docker export` to export the container rootfs for `runc`, it will cause
error if setup `cwd` in `config.json` and the `cwd` is the same with what
we specified in `docker create` with `-w`.

Signed-off-by: Lei Jitang leijitang@huawei.com

cpuguy83 commented:
Is this not needed for both Windows and Linux?

calavera commented:
yes, this should be for both, linux and windows.

Thanks @coolljt0725.

duglin commented:
@coolljt0725 would it make sense to add a test for this? Something like:
- create a container with `-w /foo`
- try `docker cp xxx:/foo .` and make sure that it doesn't generate a file not found type of error, like it would today?

coolljt0725 commented:
@cpuguy83  I thought this is needed for both Linux and Windows at first, but I found #17671 remove the `Mount` on container creation which is need for `SetupWorkingDirectory`, so at this pr, I just make it work for Linux, I think we should ping some Windows guys to have a look.

coolljt0725 committed:
Create the working directory on container creation

if create a container with -w to specify the working directory and
the directory does not exist in the container rootfs, the directory
will be created until the container start. It make docker export of
a created container and a running container inconsistent.

Signed-off-by: Lei Jitang <leijitang@huawei.com>

coolljt0725 commented:
@dugin add a test

runcom commented:
LGTM

test is actually checking the error:
```
Error response from daemon: lstat /var/lib/docker/overlay/94c8d887414fb84d98c386b2a87bfebf85697b82c597a530b4f36268a3f34cef/merged/home/foo/bar: no such file or directory
```
thanks @coolljt0725

calavera commented:
LGTM

jhowardmsft commented:
Ah, no, this is a bad fix. The failing test on Windows is real and is now affecting all Windows PRs. Working on a fix...
